### ISSUE TITLE
Unattended Build Agent guidance regarding PAT

### ISSUE THREAD

**Issue by VictorioBerra** [microsoft/artifacts-credprovider#170]
> ### Unattended Build Agent
> If you're running the command as part of an automated build on an unattended build agent, you can supply an access token directly using the VSS_NUGET_EXTERNAL_FEED_ENDPOINTS environment variable. **The use of Personal Access Tokens is recommended.**

PATs are from my personal Azure account right? Is there such thing as a PAT for a service account? I do not follow how a build agent used by many devs should rely on the token from a single devs account. What if they leave or are terminated? Are there any alternatives to this?

**Comment by aasim** [comment]
There are predefined variables in the build system that you can use as a PAT, please read this doc for more info: https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#systemaccesstoken, the $System.AccessToken variable carries the PAT for the running build identity.

**Comment by VictorioBerra** [comment]
I am more speaking about deploying from on-prem build agents that are not running azure's agent. Deploying from a windows VM to an IIS server on another windows VM for example. In order to go get stuff from artifacts you need a token, is there a way to use something other than a PAT? Like a SAT? Service Account Token?

**Comment by aasim** [comment]
We have a new [NuGetAuthenticate](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/package/nuget-authenticate?view=azure-devops) task, which is used to authenticate a build agent for all feeds in a given collection (or using service connections for feeds in different collections). You won't have to specify PATs etc. for this, we automatically take care of authentication using our credential provider. 

This task is currently only available on hosted agents, however, we have a script that we have developed that will install this task on your on-prem agents as well. Would you like to learn more about this approach? I can help you through the installation process for the same. 

As for your other question, the $System.AccessToken is the token for the build identity, it is therefore not related to a single individual's PAT, and can be shared by anyone whom you have given access.

**Comment by VictorioBerra** [comment]
@aasim I would be interested in hearing more. This requires an Azure Pipelines Agent to be downloaded and installed on my VM correct? I think this is the disconnect here. I am using an Atlassian Bamboo agent and I would like `dotnet restore` to download my artifacts without the use of a PAT since a PAT is tied to an individual user account, expires in 90 days and is not a service account.

Things like MyGet support this: https://docs.myget.org/docs/reference/feed-endpoints#Private_feed_endpoints_and_authentication

If the answer is simply that I either us a PAT, or an Azure Pipelines Agent

**Comment by aasim** [comment]
Ah! now I understand. Yes, for your case, a solution is to either use PATs or use the Azure pipelines agent with our authenticate task.

**Comment by polanfong** [comment]
So I take it that unattended non-Azure build agents will have to rely on some dev's PAT and have to update it periodically when that PAT expires, as stated in @VictorioBerra 's initial post?

**Comment by rythos42** [comment]
This issue sounds similar to how I'd like to use this. My use-case is that I have an unattended Jenkins build that wishes to read from an Azure Artifacts NuGet feed. 

I don't like that it is required to authenticate using a specific developers account, and would love to be able to use an SSH key or API key of some kind. 

Is some feature that supports this flow on the roadmap?

**Comment by LarsPellarin** [comment]
I had a "classic" Azure DevOps pipeline with a `nuget  update` task followed by a powershell script invoking a REST call to DevOps' `/commits` . 
We converted this to `dotnet` based yaml scripts.

The clasic pipeline used to run fine until I switched to `DotNetCoreCLI@2` tasks and the new credentials provider on our on premise agent. That is, all the `DotNetCoreCLI@2` tasks authenticate just fine, but the powershell script cant authenticate with the `$env:SYSTEM_ACCESSTOKEN` anymore. It seems te be gone.

The task before the powershell task is a `dotnet add package ... --no-restore` in an attempt to avoid the credentials to be cleared from my `nuget.config`. They seems to be gone though.

I post this here as it seems related, but let me know if you prefer a separate issue.
